File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/base/BaseDialogBuilder.java
package com.github.RooneyAndShadows.lightbulb.dialogs.base;

import androidx.fragment.app.FragmentManager;

public abstract class BaseDialogBuilder<DialogType extends BaseDialogFragment> {
    protected final String dialogTag;
    protected final FragmentManager fragmentManager;
    protected String title;
    protected String message;
    protected BaseDialogFragment.DialogButtonConfiguration positiveButtonConfiguration;
    protected BaseDialogFragment.DialogButtonConfiguration negativeButtonConfiguration;
    protected BaseDialogFragment.DialogButtonClickListener onPositiveClickListener;
    protected BaseDialogFragment.DialogButtonClickListener onNegativeClickListener;
    protected BaseDialogFragment.DialogShowListener onShowListener;
    protected BaseDialogFragment.DialogHideListener onHideListener;
    protected BaseDialogFragment.DialogCancelListener onCancelListener;
    protected BaseDialogFragment.DialogAnimationTypes animation;
    protected BaseDialogFragment.DialogTypes dialogType;
    protected boolean cancelableOnClickOutside = true;

    public BaseDialogBuilder(FragmentManager manager, String dialogTag) {
        this.dialogTag = dialogTag;
        this.fragmentManager = manager;
    }

    public BaseDialogBuilder<DialogType> withTitle(String title) {
        this.title = title;
        return this;
    }

    public BaseDialogBuilder<DialogType> withMessage(String message) {
        this.message = message;
        return this;
    }

    public BaseDialogBuilder<DialogType> withPositiveButton(BaseDialogFragment.DialogButtonConfiguration positiveButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        this.positiveButtonConfiguration = positiveButtonConfiguration;
        this.onPositiveClickListener = onClickListener;
        return this;
    }

    public BaseDialogBuilder<DialogType> withNegativeButton(BaseDialogFragment.DialogButtonConfiguration negativeButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        this.negativeButtonConfiguration = negativeButtonConfiguration;
        this.onNegativeClickListener = onClickListener;
        return this;
    }

    public BaseDialogBuilder<DialogType> withOnCancelListener(BaseDialogFragment.DialogCancelListener listener) {
        this.onCancelListener = listener;
        return this;
    }

    public BaseDialogBuilder<DialogType> withOnShowListener(BaseDialogFragment.DialogShowListener listener) {
        this.onShowListener = listener;
        return this;
    }

    public BaseDialogBuilder<DialogType> withOnHideListener(BaseDialogFragment.DialogHideListener listener) {
        this.onHideListener = listener;
        return this;
    }

    public BaseDialogBuilder<DialogType> withCancelOnClickOutsude(boolean closeOnClickOutside) {
        this.cancelableOnClickOutside = closeOnClickOutside;
        return this;
    }

    public BaseDialogBuilder<DialogType> withDialogType(BaseDialogFragment.DialogTypes dialogType) {
        this.dialogType = dialogType;
        return this;
    }

    public BaseDialogBuilder<DialogType> withAnimations(BaseDialogFragment.DialogAnimationTypes animation) {
        this.animation = animation;
        return this;
    }

    public abstract DialogType buildDialog();
}

File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/base/BaseDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.base

import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LifecycleOwner
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.*
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.callbacks.*

abstract class BaseDialogBuilder<DialogType : BaseDialogFragment?> @JvmOverloads constructor(
    protected val dialogLifecycleOwner: LifecycleOwner? = null,
    protected val fragmentManager: FragmentManager,
    protected val dialogTag: String
) {
    protected var title: String? = null
    protected var message: String? = null
    protected var positiveButtonConfiguration: DialogButtonConfiguration? = null
    protected var negativeButtonConfiguration: DialogButtonConfiguration? = null
    protected var onPositiveClickListener: DialogButtonClickListener? = null
    protected var onNegativeClickListener: DialogButtonClickListener? = null
    protected var onShowListener: DialogShowListener? = null
    protected var onHideListener: DialogHideListener? = null
    protected var onCancelListener: DialogCancelListener? = null
    protected var animation: DialogAnimationTypes? = null
    protected var dialogType: DialogTypes? = null
    protected var dialogListeners: DialogListeners? = null
    protected var cancelableOnClickOutside = true

    open fun withTitle(title: String?): BaseDialogBuilder<DialogType> {
        this.title = title
        return this
    }

    open fun withMessage(message: String?): BaseDialogBuilder<DialogType> {
        this.message = message
        return this
    }

    open fun withPositiveButton(
        configuration: DialogButtonConfiguration?,
        onClickListener: DialogButtonClickListener?
    ): BaseDialogBuilder<DialogType> {
        this.positiveButtonConfiguration = configuration
        onPositiveClickListener = onClickListener
        return this
    }

    open fun withNegativeButton(
        configuration: DialogButtonConfiguration?,
        onClickListener: DialogButtonClickListener?
    ): BaseDialogBuilder<DialogType> {
        this.negativeButtonConfiguration = configuration
        onNegativeClickListener = onClickListener
        return this
    }

    open fun withOnCancelListener(listener: DialogCancelListener?): BaseDialogBuilder<DialogType> {
        onCancelListener = listener
        return this
    }

    open fun withOnShowListener(listener: DialogShowListener?): BaseDialogBuilder<DialogType> {
        onShowListener = listener
        return this
    }

    open fun withOnHideListener(listener: DialogHideListener?): BaseDialogBuilder<DialogType> {
        onHideListener = listener
        return this
    }

    open fun withCancelOnClickOutside(closeOnClickOutside: Boolean): BaseDialogBuilder<DialogType> {
        cancelableOnClickOutside = closeOnClickOutside
        return this
    }

    open fun withDialogType(dialogType: DialogTypes?): BaseDialogBuilder<DialogType> {
        this.dialogType = dialogType
        return this
    }

    open fun withAnimations(animation: DialogAnimationTypes?): BaseDialogBuilder<DialogType> {
        this.animation = animation
        return this
    }

    open fun withDialogListeners(listeners: DialogListeners?): BaseDialogBuilder<DialogType> {
        dialogListeners = listeners
        return this
    }

    abstract fun buildDialog(): DialogType
}